src/core/tsi/ssl_transport_security.cc

```diff
@@ -1026,28 +1025,137 @@ static int GetCrlFromProvider(X509_STORE_CTX* ctx, X509_CRL** crl_out,
+  if (!valid) {
+    return valid;
+  }
+  // 6.3.3c Not supporting deltas
+  // 6.3.3d Not supporting reasons masks
+  // 6.3.3e Not supporting reasons masks
+  // 6.3.3f We only support direct CRLs so these paths are by definition the
+  // same.
+  // 6.3.3g Verify CRL Signature
+  valid = grpc_core::VerifyCrlSignature(crl, issuer);
+  return valid;
+}
+
+// Check if a given certificate is revoked
+// Returns 1 if the certificate is not revoked, 0 if the certificate is revoked
+static int CheckCertRevocation(grpc_core::experimental::CrlProvider* provider,
+                               X509* cert, X509* issuer) {
+  auto crl = GetCrlFromProvider(provider, cert);
+  // Not finding a CRL is a specific behavior. Per RFC5280, not having a CRL to
+  // check for a given certificate means that we cannot know for certain if the
+  // status is Revoked or Unrevoked and instead if Undetermined. How a user
+  // handles an Undetermined CRL is up to them. We use absl::IsNotFound as an
+  // analogue for not finding the Crl from the provider, thus the certificate in
+  // question is undetermined.
+  if (absl::IsNotFound(crl.status())) {
+    // TODO(gtcooke94) knob for undetermined being revoked or unrevoked. By
+    // default, unrevoked.
+    return 1;
+  } else if (!crl.ok()) {
+    // This is an unexpected error, return false
+    return 0;
+  }
+  // Validate the crl
+  // RFC5280 6.3.3(a-i)
+  if (!ValidateCrl(cert, issuer, *crl)) {
+    X509_CRL_free(*crl);
+    return 0;
+  }
+
+  // RFC5280 6.3.3j Actually check revocation
+  // Look for serial number of certificate in CRL  X509_REVOKED* rev =
+  // nullptr;
+  X509_REVOKED* rev;
+  if (X509_CRL_get0_by_cert(*crl, &rev, cert)) {
+    // cert is revoked
+    X509_CRL_free(*crl);
+    return 0;
+  }
+  // The certificate is not revoked
+  // RFC5280k - Not supported
+  // RFC5280l - Not supported
+  X509_CRL_free(*crl);
   return 1;
 }
 
-// When using CRL Providers, this function used to override the default
-// `check_crl` function in OpenSSL using `X509_STORE_set_check_crl`.
-// CrlProviders put the onus on the users to provide the CRLs that they want to
-// provide, and because we override default CRL fetching behavior, we can expect
-// some of these verification checks to fails for custom CRL providers as well.
-// Thus, we need a passthrough to indicate to OpenSSL that we've provided a CRL
-// and we are good with it.
-static int CheckCrlPassthrough(X509_STORE_CTX* /*ctx*/, X509_CRL* /*crl*/) {
+// Checks each certificate in the chain for revocation
+// returns 0 if any cert in the chain is revoked, 1 otherwise.
+static int CheckChainRevocation(
+    X509_STORE_CTX* ctx, grpc_core::experimental::CrlProvider* provider) {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000
+  STACK_OF(X509)* chain = X509_STORE_CTX_get0_chain(ctx);
+#else
+  STACK_OF(X509)* chain = X509_STORE_CTX_get_chain(ctx);
+#endif
+  if (chain == nullptr) {
+    return 0;
+  }
+  size_t chain_length = sk_X509_num(chain);
```

Comment:
sk_x509_num() returns an int, which can be -1, and you're about to jam it into a size_t which is unsigned.
it can return -1 on failure, which includes if the pointer is NULL. Rather than worry about this at all, I would
simply remove the superfluous check for chain being nullptr above, make chain_length and int, and check
for chain_length <= 0 below.

Reply:
I think I pulled the `size_t` from seeing how the fn is used in the boringssl codebase (https://github.com/google/boringssl/blob/1a118bbf76ab52698961a1c7ec717ab0080b037d/crypto/x509/x509_cmp.c#L201), and in my IDE my type hinting says it returns a `size_t`. I think I'm not quite clear on the return, could you help me understand? 
[here](https://github.com/google/boringssl/blob/1a118bbf76ab52698961a1c7ec717ab0080b037d/include/openssl/stack.h#L296) it looks like the sk_TYPE_num returns a `size_t`, but [here](https://github.com/openssl/openssl/blob/74fd6823884e27c18ec3fe7bd99b9bc02e6f31f3/include/openssl/stack.h#L29) it appears to return an `int`? I'm not sure if this could be some weirdness in how grpc's bazel dependencies are pulling BoringSSL?

Just want to make sure I get it right

Reply:
They are different in OpenSSL vs BoringSSL,  in BoringSSL they are size_t, in OpenSSL they are int.

Reply:
Discussed offline, we are using `size_t` and then also checking against a max size.

Reply:
ack. lgtm